### PR TITLE
Fix GradedModularFormElement multiplication

### LINKED ISSUE #35932: Non-homogeneous GradedModularFormElement multiplication and exponentiation is incorrect
### Steps To Reproduce

Code:

```python
sage: M = ModularFormsRing(1)
sage: E4, E6 = M.gens()
sage: E4 + M(1)
2 + 240*q + 2160*q^2 + 6720*q^3 + 17520*q^4 + 30240*q^5 + O(q^6)
sage: (E4 + M(1))^2
3 + 720*q + 64080*q^2 + 1056960*q^3 + 7943760*q^4 + 37530720*q^5 + O(q^6)
```

### Expected Behavior

Well, it should start with `4 + ...` since `E4 + M(1)` start with `2 + ...`. In particular, it should be `E4^2 + 2 * E4 + 1` instead of the current computed `E4^2 + E4 + 1`.

### Actual Behavior

I found the buggy line, so I will just report it (and make a PR soon): On line 3558 of `sage.modular.modform`, i.e. inside the `_mul_` method of `GradedModularFormElement`, it has the following incorrect line:

```python
        GM = self.__class__
        f_self = self._forms_dictionary
        f_other = other._forms_dictionary
        f_mul = {}
        for k_self in f_self.keys():
            for k_other in f_other.keys():
                f_mul[k_self + k_other] = f_self[k_self]*f_other[k_other]        # Line 3558
        return GM(self.parent(), f_mul)
```
It should be `+=` instead of `=`, just like normal polynomial multiplication.

### Additional Information

_No response_

### Environment

```markdown
- **Sage Version**: 10.0
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by grhkm21**
Fix #35932 - incorrect multiplication for GradedModularFormElement.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by grhkm21** [comment]
@DavidAyotte Hi, this is the new issue that I created for fixing multiplication. Took me a while to figure out how to get git working (rebase/stash/branch checkout/...). It should be ready

File: src/sage/modular/modform/element.py
Comment (DavidAyotte) [inline src/sage/modular/modform/element.py]: "trac" refers to the old issues tracker of SageMath. The following is more up to date:
```suggestion
        This shows that the issue at :issue:`35932` is fixed::
```

File: src/sage/modular/modform/element.py
Comment (DavidAyotte) [inline src/sage/modular/modform/element.py]: See comment below
```suggestion
```

File: src/sage/modular/modform/element.py
Comment (DavidAyotte) [inline src/sage/modular/modform/element.py]: I think this would be more efficient since it don't require the creation of a new parent `M` and changing the base ring.
```suggestion
                try:
                     f_mul[k_self + k_other] += f_self[k_self]*f_other[k_other]
                except KeyError:
                     f_mul[k_self + k_other] = f_self[k_self]*f_other[k_other]
```

**Review by DavidAyotte — CHANGES_REQUESTED** [review]
Thanks for the fix! Overall it looks good. It should be good to go once the comments are applied and the Github actions have run.

**Comment by grhkm21** [comment]
I have done that, thanks a lot

**Comment by DavidAyotte** [comment]
Can a maintainer accept the workflow runs for this Pull Request? @mkoeppe ? Thanks in advance.

File: src/sage/modular/modform/element.py
Comment (mkoeppe) [inline src/sage/modular/modform/element.py]: try https://docs.python.org/3/library/collections.html#collections.defaultdict

Comment (grhkm21) [reply src/sage/modular/modform/element.py]: Thanks for the feedback! I have done that, but I am not sure if I have to cast back to `dict` at the end or if it is fine to leave it as `defaultdict`.

Comment (mkoeppe) [reply src/sage/modular/modform/element.py]: It's fine to leave it as is. The module element constructor accepts it just fine

File: src/sage/modular/modform/element.py
Comment (DavidAyotte) [inline src/sage/modular/modform/element.py]: I know that it isn't from your changes, but this commented line looks strange to me. It looks like someone was trying something but forgot to delete it and it somehow got merged. If you don't mind you can delete it.

Comment (grhkm21) [reply src/sage/modular/modform/element.py]: I think me or my editor might have done it accidentally, since it gives some linting error. I have reverted the change now

**Review by DavidAyotte — APPROVED** [review]

**Pull request merged**
